hibernate-core/src/main/java/org/hibernate/id/uuid/UuidVersion7Strategy.java

```diff
@@ -0,0 +1,123 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.id.UUIDGenerationStrategy;
+
+import static java.time.Instant.EPOCH;
+import static java.time.temporal.ChronoUnit.MILLIS;
+
+/**
+ * Implements UUID Version 7 generation strategy as defined by the <a href="https://datatracker.ietf.org/doc/html/rfc9562#name-uuid-version-7">RFC 9562</a>.
+ *
+ * <ul>
+ *     <li>48 bits - 48-bit big-endian unsigned number of the Unix Epoch timestamp in milliseconds.</li>
+ *     <li>4 bits - version field, set to 0b0111 (7).</li>
+ *     <li>
+ *         12 bits - sub-milliseconds part of timestamp (resolution approximately 1/4 of millisecond)
+ *         to guarantee additional monotonicity.
+ *     </li>
+ *     <li>2 bits - variant field, set to 0b10.</li>
+ *     <li>14 bits - counter to guarantee additional monotonicity, resets to 0 when timestamp changes. </li>
+ *     <li>48 bits - pseudorandom data to provide uniqueness.</li>
+ * </ul>
+ *
+ * @author Cedomir Igaly
+ */
+public class UuidVersion7Strategy implements UUIDGenerationStrategy, UuidValueGenerator {
+
+	public static final UuidVersion7Strategy INSTANCE = new UuidVersion7Strategy();
+
+	private static class Holder {
+
+		static final SecureRandom numberGenerator = new SecureRandom();
+	}
+
+	private final Lock lock = new ReentrantLock( true );
+
+	private Duration currentTimestamp;
+
+	private final AtomicLong clockSequence;
+
+	public UuidVersion7Strategy() {
+		this( getCurrentTimestamp(), 0 );
+	}
+
+	public UuidVersion7Strategy(final Duration currentTimestamp, final long clockSequence) {
+		this.currentTimestamp = currentTimestamp;
+		this.clockSequence = new AtomicLong( clockSequence );
+	}
+
+	/**
+	 * A variant 7
+	 */
+	@Override
+	public int getGeneratedVersion() {
+		/*
+		 *	UUIDv7 features a time-ordered value field derived from the widely implemented and well-
+		 * known Unix Epoch timestamp source, the number of milliseconds since midnight 1 Jan 1970 UTC,
+		 * leap seconds excluded.
+		 */
+		return 7;
+	}
+
+	/**
+	 * Delegates to {@link #generateUuid}
+	 */
+	@Override
+	public UUID generateUUID(SharedSessionContractImplementor session) {
+		return generateUuid( session );
+	}
+
+	/**
+	 * @param session session
+	 *
+	 * @return UUID version 7
+	 * @see UuidValueGenerator#generateUuid(SharedSessionContractImplementor)
+	 */
+	@Override
+	public UUID generateUuid(SharedSessionContractImplementor session) {
+		final Duration currentTimestamp = getCurrentTimestamp();
+
+		final long seq = getSequence( currentTimestamp );
+
+		final long millis = currentTimestamp.getSeconds() * 1000 + currentTimestamp.getNano() / 1_000_000;
+		final long nanosPart = Math.round( ( currentTimestamp.getNano() % 1_000_000L ) * 0.004096 );
+
+		return new UUID(
+				millis << 16 & 0xFFFF_FFFF_FFFF_0000L | 0x7000L | nanosPart & 0xFFFL,
```

Comment:
Feel fine to ignore this comment, but would it make it more self documenting to save this mask to a constant?

Reply:
I would advise against code comments, as code comments tend to trend to lie as the project ages. 
https://dev.to/ruben_alapont/magic-numbers-and-magic-strings-its-time-to-talk-about-it-ci2